src/Verifier/SAW/Translation/Coq/Term.hs

```diff
@@ -289,11 +289,9 @@ asApplyAllRecognizer t = do _ <- asApp t
 -- restoring the current environment before returning.
 withLocalLocalEnvironment :: TermTranslationMonad m => m a -> m a
 withLocalLocalEnvironment action = do
-  env <- view localEnvironment <$> get
-  used <- view unavailableIdents <$> get
+  s <- get
   result <- action
-  modify $ set localEnvironment env
-  modify $ set unavailableIdents used
+  put s
```

Comment:
What about the `globalDeclarations` and `localDeclarations` fields; if `action` updates either of those, we don't want to throw away the changes, do we?

Reply:
Actually, propagating changes to `localDeclarations` is exactly what was causing the errors that this change fixed: if you introduce a let-binding in a local scope, its variables are no longer usable outside that local scope. As for `globalDeclarations`, I checked, and there is no place that it is modified inside a `withLocalLocalEnvironment`.

More generally, this goes back to my suggestion somewhere else that this should really be a `Reader` and not a `State` monad: `withLocalLocalEnvironment` should really just be thought of as a `local` in a reader monad.

Reply:
My point is that the translation monad really needs to be a `Reader` + `Output` monad, and the `localDeclarations` field should be in the `Output` part, and not the `Reader` part. The `localDeclarations` field is used to keep track of all the `Constant` AST nodes that are encountered during translation, because each of them will need to be lifted out to a separate top-level Coq declaration at the end. If you treat them as part of the `Reader` state, and reset them after `withLocalLocalEnvironment`, then any `Constant` AST nodes that appear under binders will be forgotten, won't they?